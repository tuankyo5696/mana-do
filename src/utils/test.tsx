import React from 'react';
import renderer from 'react-test-renderer';

interface Config {
  props: object;
  description: string;
}

const singleSnapTest = (tree: any, description: string) => {
  test(description, () => {
    expect(tree).toMatchSnapshot();
  });
};

const testSnapshots = (Component: any, configs: Config[]) => {
  describe('snapshots', () => {
    configs.forEach((config) => {
      const { props, description } = config;
      const tree = renderer.create(<Component {...props} />).toJSON() 
      singleSnapTest(tree, description);
    });
  });
};

export { testSnapshots };
