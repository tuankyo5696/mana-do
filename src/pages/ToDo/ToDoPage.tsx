import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { useTranslation } from 'react-i18next';

import reducer, { initialState } from '../../store/reducer';
import {
  setTodos,
  createTodo,
  deleteTodo,
  toggleAllTodos,
  deleteAllTodos,
  updateTodoStatus,
  updateTodo
} from '../../store/actions';
import Service from '../../service';
import { TodoStatus } from '../../models/todo';
import { isTodoCompleted } from '../../utils';
import Button from '../../component/Button/Button';
import Input from '../../component/Input/Input';
import { notify } from '../../component/Toast/Toast';

type EnhanceTodoStatus = TodoStatus | 'ALL';

interface ITempValue {
  todoId: string;
  content: string;
}

const ToDoPage = () => {
  const getShowing: EnhanceTodoStatus = JSON.parse(localStorage.getItem('showing') || '{}') || 'ALL';

  const [{ todos }, dispatch] = React.useReducer(reducer, initialState);
  const [showing, setShowing] = useState<EnhanceTodoStatus>(getShowing);
  const [toggle, setToggle] = useState<string>('');
  const history = useHistory();
  const [isHasBackspace, setHasBackspace] = useState<boolean>(false);
  const [tempValue, setTempValue] = useState<ITempValue>({
    todoId: '',
    content: ''
  });
  const inputRef = React.useRef<HTMLInputElement>(null);
  const { t } = useTranslation();

  useEffect(() => {
    (async () => {
      const resp = await Service.getTodos();
      dispatch(setTodos(resp || []));
    })();
  }, []);

  useEffect(() => {
    if (todos.length > 0) {
      localStorage.setItem('todos', JSON.stringify(todos));
    } else {
      localStorage.removeItem('todos');
    }
    // eslint-disable-next-line
  }, [todos.length]);

  const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter' && inputRef.current) {
      try {
        if (inputRef.current.value === '') {
          notify.error(t('ERRORS.FIELD_REQUIRED'));
        } else {
          const resp = await Service.createTodo(inputRef.current.value);
          if (resp) {
            notify.success(t('INPUT.ADD_TODO_SUCCESSFULLY', { value: resp.content }));
            dispatch(createTodo(resp));
          }

          inputRef.current.value = '';
        }
      } catch (e) {
        if (e.response?.status === 401) {
          history.push('/');
        }
      }
    }
  };

  const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
    dispatch(updateTodoStatus(todoId, e.target.checked));
  };

  const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(toggleAllTodos(e.target.checked));
  };

  const onDeleteAllTodo = () => {
    notify.success(t('INPUT.REMOVE_ALL_SUCCESSFULLY'));
    dispatch(deleteAllTodos());
  };

  const showTodos = todos.filter((todo) => {
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      default:
        return true;
    }
  });
  const activeTodos = todos.reduce(function (accum, todo) {
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0);

  const renderCheckBox =
    todos.length > 0 ? (
      <Input type="checkbox" checked={activeTodos === 0} onChange={onToggleAllTodo} title="All Todo is checked" />
    ) : (
      <div />
    );

  const onFocus = () => {
    setHasBackspace(false);
  };
  const onBlur = () => {
    setToggle('');
    setTempValue({ todoId: '', content: '' });
  };
  const onDoubleClick = (todoId: string) => () => {
    setToggle(todoId);
  };

  const onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Backspace') {
      setHasBackspace(true);
    }
  };

  const onUpdateTodo = (e: React.KeyboardEvent<HTMLInputElement>, todoId: string) => {
    if (e.key === 'Enter') {
      if (tempValue.content === '' && isHasBackspace) {
        dispatch(deleteTodo(todoId));
      }
      tempValue.content && dispatch(updateTodo(todoId, tempValue.content));
      setTempValue({ todoId: '', content: '' });
      setToggle('');
    }
  };
  const onChange = (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
    setTempValue({ todoId, content: e.target.value });
  };

  const onShowing = (value: EnhanceTodoStatus) => () => {
    localStorage.setItem('showing', JSON.stringify(value));
    setShowing(value);
  };
  const renderToDoList = showTodos.map((todo, index) => {
    const { id, content } = todo;
    return (
      <div key={id + index} className="ToDo__item">
        <Input type="checkbox" checked={isTodoCompleted(todo)} onChange={(e) => onUpdateTodoStatus(e, id)} />
        {toggle === id ? (
          <Input
            type="text"
            value={id === tempValue.todoId ? tempValue.content : content}
            className="ToDo__edit"
            onChange={(e) => onChange(e, id)}
            onBlur={onBlur}
            onKeyPress={(e) => onUpdateTodo(e, id)}
            onKeyDown={onKeyDown}
            onFocus={onFocus}
            autoFocus
          />
        ) : (
          <span onDoubleClick={onDoubleClick(id)}>{content}</span>
        )}
        <Button className="Todo__delete" onClick={() => dispatch(deleteTodo(id))} children="X" />
      </div>
    );
  });

  return (
    <div className="ToDo__container">
      <h2>{t('COMMONS.TITLE_LIST_TODO')}</h2>
      <div className="Todo__creation">
        <Input
          ref={inputRef}
          className="Todo__input"
          data-testid="input_create"
          placeholder={t('INPUT.PLACEHOLDER_CREATE_TODO')}
          onKeyPress={onCreateTodo}
          title={t('INPUT.TITLE_CREATE_TODO')}
          required
        />
      </div>
      <div className="ToDo__list">{renderToDoList}</div>
      <div className="Todo__toolbar">
        {renderCheckBox}
        <div className="Todo__tabs">
          <Button
            className="Action__btn"
            onClick={onShowing('ALL')}
            children={t('All')}
            disabled={showing === 'ALL'}
            title={t('BUTTON.TITLE_ALL')}
          />
          <Button
            className="Action__btn"
            onClick={onShowing(TodoStatus.ACTIVE)}
            children={t('COMMONS.ACTIVE')}
            disabled={showing === TodoStatus.ACTIVE}
            title={t('BUTTON.TITLE_ACTIVE')}
          />
          <Button
            className="Action__btn"
            onClick={onShowing(TodoStatus.COMPLETED)}
            children={t('COMMONS.COMPLETED')}
            disabled={showing === TodoStatus.COMPLETED}
            title={t('BUTTON.TITLE_COMPLETED')}
          />
          <Button
            className="Action__btn"
            onClick={onDeleteAllTodo}
            children={t('COMMONS.CLEAR_ALL')}
            disabled={todos.length === 0}
            title={t('BUTTON.TITLE_CLEAR_ALL')}
          />
        </div>
      </div>
    </div>
  );
};

export default ToDoPage;
