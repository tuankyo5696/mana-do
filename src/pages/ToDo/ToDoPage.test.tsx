import React, { useReducer } from 'react';
import { testSnapshots } from '../../utils/test';
import { renderHook } from '@testing-library/react-hooks';
import { render, fireEvent, screen, getByTestId, act } from '@testing-library/react';
import reducer, { initialState } from '../../store/reducer';
import { TodoStatus } from '../../models/todo';
import TodoPage from './ToDoPage';
import Service from '../../service';
import ToDoPage from './ToDoPage';

describe('render To do page', () => {
  testSnapshots(TodoPage, [
    {
      props: {},
      description: 'render deafult'
    }
  ]);
});

describe('Test logic', () => {
  test('Create new todo', () => {
    jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: { value: 'abc' } });
    const newTodo = {
      content: 'abc',
      created_date: new Date().toISOString(),
      status: TodoStatus.ACTIVE,
      id: '123',
      user_id: 'firstUser'
    };
    Service.createTodo = jest.fn().mockResolvedValue(newTodo);

    const { container } = render(<ToDoPage />);
    const input = getByTestId(container, 'input_create');
    act(async () => {
      const { result, waitForNextUpdate } = renderHook(() => useReducer(reducer, initialState));
      const [state, dispatch] = result.current;
      fireEvent.change(input, { target: { value: 'abc' } });
      fireEvent.keyPress(input, { key: 'Enter', keyCode: 13 });
      dispatch({ type: 'CREATE_TODO', payload: newTodo });
      await waitForNextUpdate();
      expect(state).toEqual({ todos: [newTodo] });
    });
  });
});
