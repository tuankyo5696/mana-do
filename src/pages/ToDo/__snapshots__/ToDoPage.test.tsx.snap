// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render To do page snapshots render deafult 1`] = `
<div
  className="ToDo__container"
>
  <h2>
    To Do List
  </h2>
  <div
    className="Todo__creation"
  >
    
    <input
      className="Todo__input"
      data-testid="input_create"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="What need to be done?"
      required={true}
      style={Object {}}
      title="Press enter to add todo"
      type="text"
    />
  </div>
  <div
    className="ToDo__list"
  />
  <div
    className="Todo__toolbar"
  >
    <div />
    <div
      className="Todo__tabs"
    >
      <button
        className="Action__btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#337ab7",
            "color": "#fff",
          }
        }
        title="Show All List Todo"
        type="button"
      >
        All
      </button>
      <button
        className="Action__btn"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#28a745",
            "color": "#fff",
          }
        }
        title="Show List Todo with status is Active"
        type="button"
      >
        Active
      </button>
      <button
        className="Action__btn"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#337ab7",
            "color": "#fff",
          }
        }
        title="Show List Todo with status is Completed"
        type="button"
      >
        Completed
      </button>
      <button
        className="Action__btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "color": "#fff",
          }
        }
        title="Remove All List Todo"
        type="button"
      >
        Clear all todos
      </button>
    </div>
  </div>
</div>
`;
